### PR TITLE
remove most functions from the Kubernetes controller interface

### PR DESCRIPTION
**Pull request by billiford**
It was bothering me that we were performing a Go anti-pattern in Go Clouddriver. Specifically, "Interface all the things". Watch here for a description of this anti-pattern: https://www.youtube.com/watch?v=ltqV6pDKZD8&t=1701s

We need to have a broader conversation about this of course. First off, we wrap all HTTP client calls in an interface which [according to Effective Go](https://golang.org/doc/effective_go#interfaces_and_types) is probably not good practice nor the purpose of an interface. *However*, providing interfaces for these packages allows for easy mocking using [counterfeiter](https://github.com/maxbrunsfeld/counterfeiter), which is something I just can't do without right now. Heck, even the [Cloudfoundry CLI](https://github.com/cloudfoundry/cli/blob/8c20e5118d7cb38f52d52b69060f6b77943dce6e/.github/CONTRIBUTING.md#development-tools) uses counterfeiter. I think it's perhaps possible we're just using it wrong.

Please check out [this discussion](https://old.reddit.com/r/golang/comments/9qnijt/what_are_some_go_anti_patterns/) on Reddit around Go Anti-patterns. Most are really simple changes we could make to this package (i.e. wrapping errors, which we are already doing in some cases) that would help make Go Clouddriver use better habits. Specifically [this](https://old.reddit.com/r/golang/comments/9qnijt/what_are_some_go_anti_patterns/e8b2iky/) comment discusses interfaces - some argue that what we're doing now is using OO practices which is preferred.

I think our use of declaring an interface is in general wrong, most articles that I can find regarding mocking HTTP clients [look something like this](https://levelup.gitconnected.com/mocking-outbound-http-calls-in-golang-9e5a044c2555), instead of our usual method that looks something like
```go
type Client interface {
  GetSomething(...)
  CreateSomething(...)
  ...
}
```
I haven't made up my mind yet. I think we need to find a good example of how to mock all these clients that does not use this Anti-pattern and do that.

Regardless, this is a start to avoiding this anti-pattern. I have removed all functions from the `kubernetes.Controller` interface except the `NewClient` function. It seemed we were just wrapping any function inside this controller interface to make it easier to test (I am 100% to blame for starting this).

There's a few pros and cons here.

Pros:
- The code is simpler
- Testing is more straightforward
- We are no longer performing an anti-pattern on the Kubernetes controller (at least not to the extent we were)
- We are no longer abstracting logic away from our tests
- Our tests are much more explicit instead of implicit and are actually acting on the objects instead of being passed into some mocked function that does nothing

Cons:
- We miss out on some test coverage as most error checks are thrown from the `ToUnstructured` function which is being called multiple times throughout a given method (for example, when adding labels and annotations to an unstructured object)

Additionally, in this PR I have
- Added a few comments to exported functions
- Changed some language in the tests to be a bit more BDD
- Made use of the Kubernetes client's `unstructured.Unstructured`'s List items instead of creating our own
- Added/removed some newlines in files for style purposes

### CODE REVIEW

**Comment by billiford** [comment]
We should list the offending anti-patterns and create issues for each to fix. Please let me know if you agree or disagree. I will create the issues and begin addressing the things we agree should be addressed (after this PR is reviewed and merged).

From [this article](https://about.sourcegraph.com/go/idiomatic-go/):
- [Tiny Package Syndrome](https://about.sourcegraph.com/go/idiomatic-go/#Tiny-Package-Syndrome): while some of our client packages are small, they are indeed tested. This doesn't personally bother me and I don't think this needs attention in this repo.
- [Premature Exportation](https://about.sourcegraph.com/go/idiomatic-go/#Premature-Exportation): this is something we need to address by moving anything *internal* to an `internal` package. EZPZ
- [Package Util](https://about.sourcegraph.com/go/idiomatic-go/#Package-util): I think we have a package util in here that has one test. Easy fix is just to move to the internal package.
- [Config Structs](https://about.sourcegraph.com/go/idiomatic-go/#Config-Structs): Check to make sure we aren't using config structs. I think we might be for our server setup. Easy fix is to add `With...` func setters.
- [Pointer all the Things](https://about.sourcegraph.com/go/idiomatic-go/#Pointer-All-The-Things): I don't think we're too bad on this one. The Kubernetes package returns *a lot* of pointers, so we might consider converting all manifests at the beginning of any Kubernetes ops stage to be of type `unstructured.Unstructured`, but that's all I can think of. Probably not worth addressing.
- [context.Value()](https://about.sourcegraph.com/go/idiomatic-go/#contextValue) - I think we do a good job letting Gin handle this and returning explicit types on our `Instance()` funcs and attaching the types to the context in our middlewares. I don't think this is worth addressing.
- [Asynchronous APIs](https://about.sourcegraph.com/go/idiomatic-go/#Asynchronous-APIs): I don't think we do this at all.
- [If-then-else](https://about.sourcegraph.com/go/idiomatic-go/#If-then-else): Let's just enable `golint` in our linting action and it will catch these.
- [Panic in a Lib](https://about.sourcegraph.com/go/idiomatic-go/#Panic-in-a-Lib): Nope.
- [Interface all the Things](https://about.sourcegraph.com/go/idiomatic-go/#Interface-All-The-Things): I think we should continue to use interfaces because they provide valuable mocking, but **create the most minimal interfaces possible and not interface anything unnecessary** (like this PR is trying to fix)
- [interface{}](https://about.sourcegraph.com/go/idiomatic-go/#interface): We hardly use interfaces IIRC, nothing to do here.

File: pkg/kubernetes/version.go
Comment (yasmin-abdullah) [inline pkg/kubernetes/version.go:94]: Yes to this. Magic numbers are bad.

File: pkg/http/core/kubernetes/deploy_test.go
Comment (billiford) [reply pkg/http/core/kubernetes/deploy_test.go:148]: The only way to trigger this error is for `ToUnstructured` to return an error. I think we should convert all the manifests to unstructured at the start and just work from there.

File: pkg/http/core/kubernetes/deploy_test.go
Comment (billiford) [reply pkg/http/core/kubernetes/deploy_test.go:239]: We can't trigger these errors anymore since the only thing that can fail is the `ToUnstructured` function.

File: pkg/http/core/kubernetes/deploy_test.go
Comment (yasmin-abdullah) [reply pkg/http/core/kubernetes/deploy_test.go:148]: Yes, That makes sense since we need to convert it to work with it.

**Review by yasmin-abdullah — APPROVED** [review]

**Comment by billiford** [comment]
@yasmin-abdullah thanks for your input!

@dmrogers7 when you have the time please add anything to the discussion you'd like, but here's the issues I'd like to open after you've reviewed and merged this PR.
- anti-pattern: move all things that should not be exported to an internal package
- anti-pattern: remove package `util`
- anti-pattern: remove any "config" structs and replace with default values, arguments to constructors, and setters
- anti-pattern: remove all non-request-scoped usages of `context.Value`
- anti-pattern: remove all panics
- enable `golint` in our linter action
- minimize all interfaces
- add contributing, license, and maintainers files

I'd love to open these issues and start addressing them in my free time, so let me know please!

**Review by dmrogers7 — APPROVED** [review]

**Pull request merged**
